sql/src/main/java/io/crate/execution/engine/join/JoinOperations.java

```diff
@@ -113,4 +125,93 @@ public static Projection createJoinProjection(List<Symbol> outputs, List<Symbol>
+                continue;
+            }
+            JoinPair prevPair = joinPairsMap.put(Sets.newHashSet(joinPair.left(), joinPair.right()), joinPair);
+            if (prevPair != null) {
+                throw new IllegalStateException("joinPairs contains duplicate: " + joinPair + " matches " + prevPair);
+            }
+        }
+        return joinPairsMap;
+    }
+
+    /**
+     * Converts any implicit join conditions of the WHERE clause to explicit {@link JoinPair}.
+     * Every join condition that gets to be converted is removed from the {@code splitQueries}
+     * @param explicitJoinPairs The explicitJoinPairs as originally written in the query
+     * @param splitQueries      The remaining queries of the WHERE clause split by involved relations
+     * @return the new list of {@link JoinPair}
+     */
+    public static List<JoinPair> convertImplicitJoinConditionsToJoinPairs(List<JoinPair> explicitJoinPairs,
+                                                                          Map<Set<QualifiedName>, Symbol> splitQueries) {
+        LongBitSet processedJoinPairs = new LongBitSet(explicitJoinPairs.size());
+        Iterator<Map.Entry<Set<QualifiedName>, Symbol>> queryIterator = splitQueries.entrySet().iterator();
+        List<JoinPair> newJoinPairs = new ArrayList<>(explicitJoinPairs.size() + splitQueries.size());
+
+        while (queryIterator.hasNext()) {
+            Map.Entry<Set<QualifiedName>, Symbol> queryEntry = queryIterator.next();
+            Set<QualifiedName> relations = queryEntry.getKey();
+
+            if (relations.size() == 2) { // If more than 2 relations are involved it cannot be converted to a JoinPair
+                Symbol implicitJoinCondition = queryEntry.getValue();
+                JoinPair newJoinPair = null;
+                for (int i = 0; i < explicitJoinPairs.size(); i++) {
+                    JoinPair joinPair = explicitJoinPairs.get(i);
+                    if (relations.contains(joinPair.left()) && relations.contains(joinPair.right())) {
+                        processedJoinPairs.set(i);
+                        // If a JoinPair with the involved relations already exists then depending on the JoinType:
+                        //  - INNER JOIN: the implicit join condition can be "AND joined" with
+                        //                the existing explicit condition of the JoinPair.
+                        //
+                        //  - CROSS JOIN: the implicit join condition becomes explicit and set on the JoinPair
+                        //                which becomes an INNER JOIN
+                        //
+                        //  - Outer Joins: Queries in the WHERE clause must be semantically applied after the outer join
+                        //                 and cannot be merge with the conditions of the ON clause.
+                        //
+                        if (joinPair.joinType() == JoinType.INNER || joinPair.joinType() == JoinType.CROSS) {
```

Comment:
Maybe we could use a `switch` statement here and have an error case for `SEMI` / `ANTI` ? (To make the valid "else" clause more explicit)

Reply:
hm, I don't really get you here. If it's not an INNER/CROSS we still need to add this existing joinPair to the list.

Reply:
I guess since the comment doesn't mention `ANTI` and `SEMI` I assumed it would be an error case.

Reply:
ah ok, yes, will add the case to the comment. thx!